### PR TITLE
coloring shapes by categorical variable

### PR DESCRIPTION
**Pull request by Sonja-Stockhaus**
This should fix multiple problems when coloring shapes, namely:
```python
# data reading, preprocessing
sdata = sd.read_zarr("visium_associated_xenium_io.zarr")
sdata.table.obs["cluster"] = "a"
sdata.table.obs.loc[sdata.table.obs.array_row==1.0, "cluster"] = "b"
sdata.table.obs.loc[sdata.table.obs.array_row==47.0, "cluster"] = "c"
sdata.table.obs.cluster = sdata.table.obs.cluster.astype("category")
``` 

1 Incorrect legend coloring when using the groups argument:
```python
sdata.pl.render_shapes('CytAssist_FFPE_Human_Breast_Cancer', color="cluster", groups=["c", "b"]).pl.show("global")
``` 
grafik

2 palette parameter in render_shapes doesn't have influence on the plot (= https://github.com/scverse/spatialdata-plot/issues/145)
```python
sdata.pl.render_shapes('CytAssist_FFPE_Human_Breast_Cancer', color="cluster", groups=["c", "b"], palette=ListedColormap(['blue', 'red'])).pl.show("global")
``` 
grafik

Plus some downstream bugfixes that came up after that

### CODE REVIEW

**Comment by timtreis** [comment]
Is this bug already solved? Or were these diagnostic plots?

**Comment by Sonja-Stockhaus** [comment]
Bug is solved, I still want to add some tests

File: src/spatialdata_plot/pl/render.py [hunk 1]
```diff
@@ -122,9 +137,12 @@ def _render_shapes(
         cax = ax.add_collection(_cax)
 
         # Using dict.fromkeys here since set returns in arbitrary order
-        palette = (
-            ListedColormap(dict.fromkeys(color_vector)) if render_params.palette is None else render_params.palette
-        )
+        # remove the color of NaN values, else it might be assigned to a category
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:140]: Better to have this comment directly next to the `dict.from_keys`

File: src/spatialdata_plot/pl/utils.py [hunk 2]
```diff
@@ -843,8 +843,9 @@ def _set_color_source_vec(
 
         if groups is not None:
             color_source_vector = color_source_vector.remove_categories(categories.difference(groups))
+            categories = groups
 
-        color_map = dict(zip(categories, _get_colors_for_categorical_obs(categories)))
+        color_map = dict(zip(categories, _get_colors_for_categorical_obs(categories, palette)))
         # color_map = _get_palette(
         #     adata=adata, cluster_key=value_to_plot, categories=categories, palette=palette, alpha=alpha
```
Comment (timtreis) [inline src/spatialdata_plot/pl/utils.py:850]: remove comment plz

File: tests/pl/test_render_shapes.py [hunk 3]
```diff
@@ -101,6 +102,26 @@ def test_plot_can_color_from_geodataframe(self, sdata_blobs: SpatialData):
...
+
+    def test_plot_can_plot_with_group_and_palette(self, sdata_blobs: SpatialData):
+        sdata_blobs.shapes["blobs_polygons"]["cluster"] = "c1"
+        sdata_blobs.shapes["blobs_polygons"].iloc[3:5, 1] = "c2"
+        sdata_blobs.shapes["blobs_polygons"]["cluster"] = sdata_blobs.shapes["blobs_polygons"]["cluster"].astype(
+            "category"
+        )
+
+        sdata_blobs.pl.render_shapes(
+            "blobs_polygons", color="cluster", groups=["c2", "c1"], palette=ListedColormap(["green", "yellow"])
```
Comment (timtreis) [inline tests/pl/test_render_shapes.py:122]: Please take care of the ListedColormap conversion inside the code, it's unintuitive to have the user feed it in.

File: CHANGELOG.md [hunk 4]
```diff
@@ -20,6 +20,7 @@ and this project adheres to [Semantic Versioning][].
 -   Multipolygons are now handled correctly (#93)
 -   Legend order is now deterministic (#143)
 -   Images no longer normalised by default (#150)
+-   Shapes can be colored by categorical variable with the `group` and `palette` arguments (#153)
```
Comment (timtreis) [inline CHANGELOG.md:23]: Please make clear that `groups` is used to _filter_ the data and `palette` is used to color the (potentially remaining) categories. Also, what happens if I add a palette argument to a continous variable?

**Comment by timtreis** [comment]
Meeting notes
- [ ] Implement also for labels using MibiTOF
- [ ] Change Changelog text
- [ ] Internalise ListedColormap conversion
- [ ] Docstrings
- [ ] one string is accepted for both arguments

File: src/spatialdata_plot/pl/basic.py [hunk 5]
```diff
@@ -183,8 +183,11 @@ def render_shapes(
             Key in :attr:`anndata.AnnData.layers` or `None` for :attr:`anndata.AnnData.X`.
         palette
             Palette for discrete annotations, see :class:`matplotlib.colors.Colormap`.
```
Comment (timtreis) [inline src/spatialdata_plot/pl/basic.py:185]: L185 and L186 disagree, no?

Comment (Sonja-Stockhaus) [reply src/spatialdata_plot/pl/basic.py:185]: because of the :class:`matplotlib.colors.Colormap`?

File: src/spatialdata_plot/pl/utils.py [hunk 6]
```diff
@@ -784,20 +788,21 @@ def _get_colors_for_categorical_obs(
...
 
-    # otherwise, single chanels turn out grey
+    # otherwise, single channels turn out grey
     color_idx = np.linspace(0, 1, len_cat) if len_cat > 1 else [0.7]
 
     if isinstance(palette, str):
-        cmap = plt.get_cmap(palette)
-        palette = [to_hex(x) for x in cmap(color_idx, alpha=alpha)]
+        # cmap = plt.get_cmap(palette)
+        # palette = [to_hex(x) for x in cmap(color_idx, alpha=alpha)]
```
Comment (timtreis) [inline src/spatialdata_plot/pl/utils.py:796]: remove comments

File: src/spatialdata_plot/pl/basic.py [hunk 7]
```diff
@@ -182,9 +182,13 @@ def render_shapes(
         layer
             Key in :attr:`anndata.AnnData.layers` or `None` for :attr:`anndata.AnnData.X`.
         palette
-            Palette for discrete annotations, see :class:`matplotlib.colors.Colormap`.
+            Palette for discrete annotations. List of valid color names that should be used
+            for the categories (all or as specified by `groups`). For a single category,
```
Comment (timtreis) [inline src/spatialdata_plot/pl/basic.py:186]: good descriptions 👌

**Review by timtreis — APPROVED** [review]

**Pull request merged**